File: C_plus_plus/REmove_Duplicates_from_Sorted_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode *last = NULL,*result = NULL;
        if(!head||!(head->next)) return head;
        while(head){
            int isDul = false;
            while(head&&head->next&&head->val==head->next->val){ 
                head = head->next;
                isDul = true;
            }
            if(!isDul){
                if(last){
                    last->next = head;
                    last = head;
                }
                else
                    last = result = head;
            }
            head = head->next;
        }
        if(last) last->next = NULL;
        return result;
    }
};

File: java/Remove_Duplicates_from_Sorted_List_II.java
// Using count to record the continuous val, and remove it if count > 1
// One pass 
// Time is O(n), space is O(1).

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode pre = null, p = head;
        while (p != null) {
            int count = 1;
            while(p.next != null && p.val == p.next.val) {
                count++;
                p = p.next;
            }
            if (count == 1) {
                if (pre == null) {
                    pre = head = p;
                }
                else {
                    pre.next = p;
                    pre = p;
                }
            }
            p = p.next;
        }
        if (pre == null) return null;
        pre.next = null;
        return head;
    }
}
